Implement cucumber tests for one-sided recovery and scanning

philipr-za opened the pull request:
## Description
This PR implements cucumber tests for recovering one-sided payments, scanning for them under normal wallet operation and also adds to the existing standard recovery test to spend the output after it has been recovered which revealed a few bugs. The PR also contains a number of bug fixes to make these tests possible.

- Console wallet will not enter TUI mode after recovery if its started in daemon mode, this crashes cucumber
- Persist “Known one-sided scripts” when a Comms Private/Public key is derived to be scanned for
- Fix the UTXO scanner when started as a service.
- Add Cucumber helpers to support the tests.

## How Has This Been Tested?
Cucumber tests provided

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [ ] I'm merging against the `development` branch.
* [x] I have squashed my commits into a single commit.

hansieodendaal commented on integration_tests/features/WalletRecovery.feature:
```diff
@@ -1,15 +1,49 @@
+    Then all nodes are at height 15
+    When I recover wallet WALLET_A into wallet WALLET_B connected to all seed nodes
+    Then wallet WALLET_A and wallet WALLET_B have the same balance
+    And I have wallet WALLET_C connected to all seed nodes
+    And I send 100000 uT from wallet WALLET_B to wallet WALLET_C at fee 100
+    When I mine 5 blocks on NODE
+    Then all nodes are at height 20
+    Then I wait for wallet WALLET_C to have at least 100000 uT
```
Nice addition

hansieodendaal commented on integration_tests/features/WalletRecovery.feature:
```diff
@@ -1,15 +1,49 @@
+    Then I wait for wallet WALLET_C to have at least 100000 uT
+
+@critical
+Scenario: Scan for one-sided payments
```
**NIT**: Maybe the name should change here as the test is more about recovery and not so much scanning. With the latter, I would expect WALLET_B to be started up again in its normal mode and being issued a command to scan for one-sided transactions.
Maybe something like `Recover one-sided payments not received yet` ?

philipr-za replied:
Sure, can change the name. In the end recovery is the same process of scanning and this is in the Recovery Feature but 🤷‍♂️

hansieodendaal replied:
Just a **NIT** though

hansieodendaal replied:
Let me mark it as such

hansieodendaal commented:
Wow, lots of effort went into this

delta1 commented:
Looks great, just a lint warning (how were you able to commit?! `npm install` should install husky with a pre-commit hook) 

```
1440:9  error  'identityInfo' is assigned a value but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)
```

philipr-za commented:
Rebased and passing locally with `npm test -- --tags "@critical and not @long-running and not @broken"` but still requires a manual `npm link tari_crypto` with the latest version of `tari_crypto`

hansieodendaal commented:
> Rebased and passing locally with `npm test -- --tags "@critical and not @long-running and not @broken"` but still requires a manual `npm link tari_crypto` with the latest version of `tari_crypto`

Yes, I omitted to change `"tari_crypto": "^0.8.0"` to `"tari_crypto": "^0.9.0"` in `package.json`, which does the trick. Fixed in consecutive PRs.

hansieodendaal commented on integration_tests/features/WalletRecovery.feature:
```diff
@@ -1,15 +1,49 @@
+    Then wallet WALLET_A and wallet WALLET_B have the same balance
+    And I have wallet WALLET_C connected to all seed nodes
+    And I send 100000 uT from wallet WALLET_B to wallet WALLET_C at fee 100
+    When I mine 5 blocks on NODE
```
We need a wait of a second or so here **before this step** (or some other method like mine one block then check heights) otherwise mining starts before the transaction is finalized, especially on a faster system. This test failed on one system and passed on the other due to the confirmations only going to count of 2; 2 blocks were mined and the next block template issued before the transaction was ready and validated in the mempool.

philipr-za replied:
hmm, perhaps I will rather put in a statement to wait for the transaction to be broadcast. I prefer that to adding in a finite wait which might not work on all setups.

philipr-za replied:
added `Then wallet WALLET_B detects all transactions are at least Broadcast`

hansieodendaal replied:
Yip, that did the trick.

philipr-za committed:
Implement cucumber tests for one-sided recovery and scanning

This PR implements cucumber tests for recovering one-sided payments, scanning for them under normal wallet operation and also adds to the existing standard recovery test to spend the output after it has been recovered which revealed a few bugs. The PR also contains a number of bug fixes to make these tests possible.

- Console wallet will not enter TUI mode after recovery if its started in daemon mode, this crashes cucumber
- Persist “Known one-sided scripts” when a Comms Private/Public key is derived to be scanned for
- Fix the UTXO scanner when started as a service.
- Add Cucumber helpers to support the tests.

hansieodendaal commented on integration_tests/features/support/steps.js:
```diff
@@ -996,21 +1001,42 @@ When(
+      5 * 1000,
+      5
+    );
+    if (!walletClient.isBalanceAtLeast(amount)) {
+      console.log("Balance not adequate!");
     }
+    consoleLogBalance(await walletClient.getBalance());
```
```suggestion
    consoleLogBalance(await walletClient.getBalance());
    if (!(await walletClient.isBalanceAtLeast(amount))) {
      console.log("Balance not adequate!");
    }
    expect(await walletClient.isBalanceAtLeast(amount)).to.equal(true);
```

hansieodendaal replied:
We need an assert here

hansieodendaal commented on integration_tests/features/support/steps.js:
```diff
@@ -996,21 +1001,42 @@ When(
+      5 * 1000,
+      5
+    );
+    if (!walletClient.isBalanceLessThan(amount)) {
+      console.log("Balance has not dropped below specified amount!");
+    }
+    consoleLogBalance(await walletClient.getBalance());
```
```suggestion
    consoleLogBalance(await walletClient.getBalance());
    if (!(await walletClient.isBalanceLessThan(amount))) {
      console.log("Balance has not dropped below specified amount!");
    }
    expect(await walletClient.isBalanceLessThan(amount)).to.equal(true);
```

hansieodendaal replied:
We need an assert here

hansieodendaal reviewed (commented):
I only reviewed the cucumber tests and have some additional comments.

Comment already addressed:
"_Mining can start before the transaction is finalized, especially on a faster system. The `Wallet recovery with connected base node staying online` test failed on one system and passed on the other due to the confirmations only going to count of 2; 2 blocks were mined and the next block template issued before the transaction was ready and validated in the mempool._"

hansieodendaal commented:
> > Rebased and passing locally with `npm test -- --tags "@critical and not @long-running and not @broken"` but still requires a manual `npm link tari_crypto` with the latest version of `tari_crypto`
> 
> Yes, I omitted to change `"tari_crypto": "^0.8.0"` to `"tari_crypto": "^0.9.0"` in `package.json`, which does the trick. Fixed in consecutive PRs.

If you change  `"tari_crypto": "^0.8.0"` to `"tari_crypto": "^0.9.0"` in `package.json` the `run-integration-tests` step should be able to pass.

stringhandler committed:
Apply suggestions from code review

Co-authored-by: Hansie Odendaal <39146854+hansieodendaal@users.noreply.github.com>
